### PR TITLE
write timestamps as long since parquet has timestamp_micros

### PR DESCRIPTION
**Pull request by robert3005**
@schlosna @pwoody @ash211 @mccheah 

will add tests shortly but wanted to run the logic by you. Looks like it's pretty straightforward with new parquet.

### CODE REVIEW

**Comment by robert3005** [comment]
Just changes Timestamp serialization to TIMESTAMP_MICROS. Need to add it to filters that parquet can handle

**Comment by ash211** [comment]
Is there an upstream ticket for this?

Also I'm concerned about backcompat -- would data written to parquet with the pre-merge version of Spark still be read correctly by Spark with this PR merged?

**Comment by robert3005** [comment]
You need a parquet release to be able to implement it. 1.9.0 hasn't come out yet. We should make a note to upstream once parquet publishes.

@ash211 why wouldn't they? Only issue is with filter pushdown

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 1]
```diff
@@ -368,14 +369,8 @@ private[parquet] class ParquetSchemaConverter(
...
       // we may resort to microsecond precision in the future.
-      //
-      // For Parquet, we plan to write all `TimestampType` value as `TIMESTAMP_MICROS`, but it's
-      // currently not implemented yet because parquet-mr 1.7.0 (the version we're currently using)
-      // hasn't implemented `TIMESTAMP_MICROS` yet.
-      //
-      // TODO Converts `TIMESTAMP_MICROS` once parquet-mr implements that.
       case TimestampType =>
-        Types.primitive(INT96, repetition).named(field.name)
+        Types.primitive(INT64, repetition).as(TIMESTAMP_MICROS).named(field.name)
```
Comment (schlosna) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: Maybe in line with @ash211 question, how does this play with `assumeInt96IsTimestamp` set from `spark.sql.parquet.int96AsTimestamp` and defaults to `true`?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java [hunk 2]
```diff
@@ -106,13 +106,12 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
       // then we need to apply the predicate push down filter
       footer = readFooter(configuration, file, range(split.getStart(), split.getEnd()));
       FilterCompat.Filter filter = getFilter(configuration);
-      ParquetFileReader reader = ParquetFileReader.open(taskAttemptContext.getConfiguration(), file, footer);
+      this.reader = ParquetFileReader.open(configuration, file, footer);
```
Comment (schlosna) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java]: if we hit this path, we'll open twice, once here and once at line 143; can we remove this and defer to opening the reader until outside this if/else?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala (hunk 1 above)
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: This code path is when writing. We just stop writing INT96. I haven't removed code path for reading INT96

**Comment by robert3005** [comment]
For anyone curious, if you write parquet pre this change and then do a filter with this change this is the exception you will see when executing the query.

```Caused by: java.lang.IllegalArgumentException: FilterPredicate column: ts's declared type (java.lang.Long) does not match the schema found in file metadata. Column ts is of type: INT96
Valid types for this column are: [class org.apache.parquet.io.api.Binary]```

**Comment by ash211** [comment]
Is that exception in the last comment final design or just an artifact of an in-progress implementation?

We definitely want to support parquet written with older versions of the various libraries to be read by newer versions of the libraries.  Otherwise we'd have to rematerialize lots of data

**Comment by pwoody** [comment]
Closing in favor of #72

**Pull request closed without merging**